### PR TITLE
Memory: sync rasterizer cache mark to fix LLE applet crash

### PR DESCRIPTION
**Pull request by wwylele**
Quick recap of how our shader cache works with the memory system: for each mapped, non-MMIO virtual memory pages, it has two states: 
 - `page_type` = `Memory` and `pointers` points to the actual memory, allowing CPU to do fast access directly
 - `page_type` = `RasterizerCachedMemory` and `pointers=nullptr`, preventing CPU from going the fast path and letting it fallback to the slow path that checks/flushes/invalidates rasterizer cache

The rasterizer toggles the page state between these two by calling `RasterizerMarkRegionCached`. However, this function only changes the page state of the current process, which would resulting inconsistent state across processes if multiple processes are running. To restore the consistency, we should do:
 - toggle cache state of ALL processes/page tables instead of only the current one
 - inherit the cache state from a global bookkeeper when the process ask to map new memory.

Performance consideration: all functions modified in this PR are in the cold path as tested, and in practice the number for registered page tables is usually 1 (normal game play), 2 (running LLE applet), or < 50 (LLE everything, ). This really shouldn't have performance impact as previously conjectured.

---
This change is [Reviewable](https://reviewable.io/reviews/citra-emu/citra/4500)

### CODE REVIEW

File: src/core/memory.cpp [hunk 1]
```diff
@@ -324,37 +375,40 @@ void MemorySystem::RasterizerMarkRegionCached(PAddr start, u32 size, bool cached
...
-                    break;
-                case PageType::RasterizerCachedMemory: {
-                    page_type = PageType::Memory;
-                    impl->current_page_table->pointers[vaddr >> PAGE_BITS] =
-                        GetPointerForRasterizerCache(vaddr & ~PAGE_MASK);
-                    break;
-                }
-                default:
-                    UNREACHABLE();
+            impl->cache_marker.Mark(vaddr, cached);
```
Comment (wwylele) [inline src/core/memory.cpp]: Tip: enable "Hide whitespace changes" in github to review this part

File: src/core/memory.cpp [hunk 2]
```diff
@@ -15,12 +15,45 @@
...
 
+class RasterizerCacheMarker {
+public:
+    void Mark(VAddr addr, bool cached) {
+        bool* p = At(addr);
+        if (p)
+            *p = cached;
+    }
+
+    bool IsCached(VAddr addr) {
```
Comment (wwylele) [inline src/core/memory.cpp]: This should be `const`, but it requires either duplicating `bool* At(VAddr addr)` or use unusual template trick to do it (which can be really surprising to a first-time reader). I chose readability, so I discarded `const`.

**Comment by tgsm** [comment]
This is what happens to Mii Maker after opening swkbd.
miimaker

**Comment by wwylele** [comment]
@tgsm I found that as well. I think it is a different bug and is pretty specific to that app, as I tested some others they didn't show the same issue.

**Comment by wwylele** [comment]
Will merge soon if no more comments

**Pull request merged**